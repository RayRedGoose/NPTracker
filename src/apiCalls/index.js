export const getData = async (type, search) => {
  const searching = (search) ? `?q=${search}`: ''
  const apiKey = 'IpOkxICOi4tJPgJfN3LXvxXuOtWHz5iZdULtD2hd'
  const url = `https://developer.nps.gov/api/v1/${type}${searching}&fields=images&limit=60&api_key=${apiKey}`

  const response = await fetch(url)
  if (!response.ok) {
    throw new Error('There was an error getting parks.')
  }

  const result = await response.json()
  return result.data
}
