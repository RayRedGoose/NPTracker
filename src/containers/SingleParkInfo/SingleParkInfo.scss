@import "../../variables.scss";
@import "../../index.scss";

.single-park {
  color: $light-color;
  padding: 20px;
  img {
    width: 30px;
    @extend %click-action;
  }
  section:first-child {
    display: grid;
    grid-template-areas: 'img head' 'img place' 'img desc' 'img foot';
    grid-template-rows: 0.75fr 0.5fr 2fr 0.75fr;
    background-color: $darken-small;
    padding: 20px;
    width: 100%;
    &>div {
      background-position: center;
      background-repeat: no-repeat;
      background-size: cover;
      grid-area: img;
      width: 40vw; height: 100%;
      margin-right: 20px;
    }
    header {
      grid-area: head;
      display: flex;
      justify-content: space-between;
      align-content: center;
      margin-bottom: 10px;
      h2 {
        color: $accent-color-primary-light;
        font-family: Salsa;
        letter-spacing: 0.05em;
      }
    }
    .location {
      grid-area: place;
      @extend %location;
      margin-top: 0;
      font-size: 18px;
    }
    .desc {
      grid-area: desc;
      margin-top: 20px;
      font-size: 14px;
      line-height: 1.5em;
      height: 130px;
      overflow: scroll;
    }
    footer {
      grid-area: foot;
      display: flex;
      align-items: flex-end;
      margin-bottom: 5px;
      p {
        cursor: pointer;
        @extend %click-action;
        display: flex;
        align-items: center;
        color: $accent-color-primary-dark;
        font-family: Salsa;
        margin-right: 40px;
        &:before {
          content: "";
          background-position: center;
          background-repeat: no-repeat;
          background-size: cover;
          width: 20px; height: 20px;
          margin-right: 10px;
        }
      }
      .bell-no-active {
        &:before {
          background-image: url('../../assets/bell.svg');
        }
      }
      .bell-active {
        &:before {
          background-image: url('../../assets/bell-full.svg');
        }
      }
      .flag-no-active {
        &:before {
          background-image: url('../../assets/done.svg');
        }
      }
      .flag-active {
        &:before {
          background-image: url('../../assets/flag-full.svg');
        }
      }
    }
  }
  section:last-child {
    margin-top: 20px;
    display: flex;
    div {
      background-color: $darken-small;
      margin-right: 20px;
      padding: 20px;
      width: 30%;
      h3 {
        color: $accent-color-primary-dark;
        font-family: Salsa;
        text-align: center;
        margin-bottom: 20px;
      }
      p {
        font-size: 14px;
        line-height: 1.5em;
        height: 230px;
        overflow: scroll;
      }
    }
  }
}
