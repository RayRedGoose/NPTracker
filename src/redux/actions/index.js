export const addProcessName = name => ({
  type: 'ADD_PROCESS_NAME',
  name
})

export const addProcessStep = step => ({
  type: 'ADD_PROCESS_STEP',
  step
})

export const increaseStep = () => ({
  type: 'INCREASE_STEP'
})

export const addUserInfo = info => ({
  type: 'ADD_USER_INFO',
  info
})

export const addParks = parks => ({
  type: 'ADD_PARKS',
  parks
})

export const changeActiveTab = id => ({
  type: 'CHANGE_ACTIVE_TAB',
  id
})
