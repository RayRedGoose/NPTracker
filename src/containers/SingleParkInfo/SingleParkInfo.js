import './SingleParkInfo.scss'
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { addItem, addItemToAll } from '_utils/localStorage'
import { selectPark, addPlannedPark, removePlannedPark } from 'redux/actions'
import { getData } from 'apiCalls'
import arrow from 'assets/more.svg'

export class SingleParkInfo extends Component {
  constructor() {
    super()
    this.state = {
      isLoaded: false
    }
  }

  componentDidMount() {
    const { type, parks } = this.props
    return (parks.length)
      ? this.findPark(type)
      : this.fetchPark(type)
  }

  findPark = (type) => {
    const { parks, selectPark } = this.props
    const search = type.replace(/-/g, ' ')
    const park = parks.find(park => park.name.toLowerCase() === search)
    selectPark(park)
    this.setState({isLoaded: true})
  }

  fetchPark = async (type) => {
    const search = type.replace(/-/g, '%20')
    try {
      const park = await getData('parks', search)
      this.props.selectPark(park[0])
      this.setState({isLoaded: true})
    }
    catch(error) {
      this.setState({error: error.message, isLoaded: false})
    }
  }

  toggleWishLish = () => {
    const { plannedParks, selectedPark } = this.props
    return (!plannedParks.includes(selectedPark.name))
      ? this.addToWishList(selectedPark.name)
      : this.removeFromWishList(selectedPark.name, plannedParks)
  }

  addToWishList = (name) => {
    this.props.addPlannedPark(name)
    addItemToAll('planning', name)
  }

  removeFromWishList = (name, plannedParks) => {
    this.props.removePlannedPark(name)
    const newPlans = plannedParks.filter(park => park !== name)
    addItem('planning', newPlans)
  }

  render() {
    const { selectedPark, plannedParks } = this.props
    const {
      fullName,
      name,
      description,
      url,
      states,
      images,
      directionsInfo,
      weatherInfo,
    } = selectedPark

    const bellType = (plannedParks.includes(name)) ? 'bell-active' : 'bell-no-active'
    const wishText = (plannedParks.includes(name)) ? 'Remove from wish list' : 'Add to wish list'

    return (
      this.state.isLoaded && <section className="single-park">
        <section>
          <div style={{backgroundImage: `url(${images[0].url})`}}></div>
          <header>
            <h2>{fullName}</h2>
            <a href={url}><img src={arrow} alt="arrow"/></a>
          </header>
          <p className="location">{states}</p>
          <p className="desc">{description}</p>
          <footer>
            <p className={ bellType } onClick={this.toggleWishLish}>{wishText}</p>
            {
              // <p className="flag-no-active">Mark as visited</p>
            }
          </footer>
        </section>
        <section>
          <div>
            <h3>Direction Info</h3>
            <p>{directionsInfo}</p>
          </div>
          <div>
            <h3>Weather Info</h3>
            <p>{weatherInfo}</p>
          </div>
        </section>
      </section>
    )
  }
}

export const mapStateToProps = ({ parks, selectedPark, plannedParks }) => ({
  parks, selectedPark, plannedParks
})

export const mapDispatchToProps = dispatch => (
  bindActionCreators({
    selectPark, addPlannedPark, removePlannedPark
  }, dispatch)
)

SingleParkInfo.propTypes = {
  parks: PropTypes.arrayOf(PropTypes.object).isRequired,
  selectedPark: PropTypes.object.isRequired,
  plannedParks: PropTypes.arrayOf(PropTypes.string).isRequired,
  selectPark: PropTypes.func.isRequired,
  addPlannedPark: PropTypes.func.isRequired,
  removePlannedPark: PropTypes.func.isRequired,
  type: PropTypes.string.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(SingleParkInfo)
