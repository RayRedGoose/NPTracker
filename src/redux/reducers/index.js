import { combineReducers } from 'redux'
import { process } from './singleReducers/process'
import { user } from './singleReducers/user'

const rootReducer = combineReducers({
  process,
  user
})

export default rootReducer
