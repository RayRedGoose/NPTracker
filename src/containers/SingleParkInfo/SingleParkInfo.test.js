import React from 'react'
import { shallow } from 'enzyme'
import { SingleParkInfo, mapStateToProps, mapDispatchToProps } from './SingleParkInfo'
import { getData } from 'apiCalls'
import { selectPark } from 'redux/actions'

jest.mock('apiCalls')

describe("SingleParkInfo", () => {
  const mockPark = {
    fullName: 'Mount Rainier National Park',
    name: 'Mount Rainier',
    description: 'Ascending to 14,410 feet above sea level, Mount Rainier stands as an icon in the Washington landscape.',
    url: 'https://www.nps.gov/mora/index.htm',
    states: 'WA',
    images: [{url: '/some.img.jpg'}],
    directionsInfo: 'Mount Rainier National Park is located in west-central Washington state.',
    weatherInfo: 'Weather patterns at Mount Rainier are strongly influenced by the Pacific Ocean and elevation.'
  }

  describe("SingleParkInfo component", () => {
    let container, instance

    const selectPark = jest.fn()

    const mockProps = {
      type: 'mount-rainier',
      parks: [mockPark],
      selectedPark: mockPark,
      selectPark
    }


    beforeEach(() => {
      container = shallow(
        <SingleParkInfo {...mockProps} />
      )

      instance = container.instance()
    })

    describe("Snapshots", () => {
      it("should match snapshot with all data passed in correctly", () => {
        container.setState({isLoaded: true})
        expect(container).toMatchSnapshot()
      })

      it("should match snapshot if there is no park data", () => {
        container.setState({isLoaded: false})
        expect(container).toMatchSnapshot()
      })
    })

    describe("componentDidMount", () => {
      it("should call findPark with type prop after rendering if there is parks data in store", () => {
        const spy = jest.spyOn(instance, 'findPark')
          .mockImplementation(jest.fn())
        instance.forceUpdate()

        instance.componentDidMount()

        expect(spy).toHaveBeenCalledWith('mount-rainier')
      })

      it("should call fetchPark with type prop after rendering if there is parks data in store", () => {
        const container = shallow(
          <SingleParkInfo {...mockProps} parks={[]} />
        )

        const instance = container.instance()

        const spy = jest.spyOn(instance, 'fetchPark')
          .mockImplementation(jest.fn())
        instance.forceUpdate()

        instance.componentDidMount()

        expect(spy).toHaveBeenCalledWith('mount-rainier')
      })
    })

    describe("findPark", () => {
      it("should call selectPark prop if findPark is called", () => {
        instance.findPark('mount-rainier')
        expect(selectPark).toHaveBeenCalledWith(mockPark)
      })

      it("should change isLoaded state to true if findPark is called", () => {
        container.setState({isLoaded: false})

        expect(container.state('isLoaded')).toEqual(false)

        instance.findPark('mount-rainier')

        expect(container.state('isLoaded')).toEqual(true)
      })
    })

    describe("fetchPark", () => {

      beforeEach(() => {
        getData.mockImplementation(() => {
          return Promise.resolve([mockPark])
        })
      })

      it("should call getData if fetchPark is called", async () => {
        await instance.fetchPark('mount-rainier')
        expect(getData).toHaveBeenCalledWith('parks', 'mount%20rainier')
      })

      it("should call selectPark prop if fetchPark is called", async () => {
        await instance.fetchPark('mount-rainier')
        expect(selectPark).toHaveBeenCalledWith(mockPark)
      })

      it("should change isLoaded state to true if fetchPark is called", async () => {
        container.setState({isLoaded: false})

        expect(container.state('isLoaded')).toEqual(false)

        await instance.fetchPark('mount-rainier')

        expect(container.state('isLoaded')).toEqual(true)
      })

      it("should change isLoaded state to false if getData is failed", async () => {
        getData.mockImplementation(() => {
          return Promise.reject(Error('Failed'))
        })

        expect(container.state('isLoaded')).toEqual(true)

        await instance.fetchPark('mount-rainier')

        expect(container.state('isLoaded')).toEqual(false)
      })

      it("should change error state to error message value if getData is failed", async () => {
        getData.mockImplementation(() => {
          return Promise.reject(Error('Failed'))
        })

        expect(container.state('error')).toEqual(undefined)

        await instance.fetchPark('mount-rainier')

        expect(container.state('error')).toEqual('Failed')
      })
    })
  })

  describe("mapStateToProps", () => {
    it("should return the object with correct keys from store", () => {
      const mockStore = {
        parks: [mockPark],
        selectedPark: mockPark,
        activeTab: 2
      }

      const expected = {
        parks: [mockPark],
        selectedPark: mockPark
      }

      const result = mapStateToProps(mockStore)

      expect(result).toEqual(expected)
    })
  })

  describe("mapDispatchToProps", () => {
    it("should call dispatch with selectPark after selectPark prop is called", () => {
      const mockDispatch = jest.fn()
      const mockProps = mapDispatchToProps(mockDispatch)
      const actionToDispatch = selectPark(mockPark)

      mockProps.selectPark(mockPark)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})
