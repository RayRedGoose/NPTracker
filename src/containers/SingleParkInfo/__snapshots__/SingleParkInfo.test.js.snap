// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleParkInfo SingleParkInfo component Snapshots should match snapshot if there is no park data 1`] = `""`;

exports[`SingleParkInfo SingleParkInfo component Snapshots should match snapshot with all data passed in correctly 1`] = `
<section
  className="single-park"
>
  <section>
    <div
      style={
        Object {
          "backgroundImage": "url(/some.img.jpg)",
        }
      }
    />
    <header>
      <h2>
        Mount Rainier National Park
      </h2>
      <a
        href="https://www.nps.gov/mora/index.htm"
      >
        <img
          alt="arrow"
          src="more.svg"
        />
      </a>
    </header>
    <p
      className="location"
    >
      WA
    </p>
    <p
      className="desc"
    >
      Ascending to 14,410 feet above sea level, Mount Rainier stands as an icon in the Washington landscape.
    </p>
    <footer>
      <p
        className="bell-no-active"
      >
        Add to wish list
      </p>
      <p
        className="flag-no-active"
      >
        Mark as visited
      </p>
    </footer>
  </section>
  <section>
    <div>
      <h3>
        Direction Info
      </h3>
      <p>
        Mount Rainier National Park is located in west-central Washington state.
      </p>
    </div>
    <div>
      <h3>
        Weather Info
      </h3>
      <p>
        Weather patterns at Mount Rainier are strongly influenced by the Pacific Ocean and elevation.
      </p>
    </div>
  </section>
</section>
`;
